readme.txt
=== Bitcoin Charts Widget ===
Plugin Name: Bitcoin Charts Widget
Plugin URI: http://www.bitcoin-welt.com/wordpress-plugin-bitcoin-charts-als-widget/
Author: Both Interact GmbH
Author URI: http://www.both-interact.com
Version: 0.4.0
Tags: bitcoin, charts, widget, bitcoin-welt, blockchain.info, service
Requires at least: 3.5.0
Tested up to: 4.2
License: GPLv2 or later
License URI: http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
Stable tag: trunk

Displays bitcoin charts as widget using data from external services such as
blockchain.info.

== Description ==

This widget displays bitcoin charts with data from different external services. 
Currently, the widget supports blockchain.info but can be extended easily.

As usual, you can adjust the widget's settings to your taste through the admin backend. 

Additionally, you can include the widget using shortcode.

= Short code =

The shortcode is [btcchart style="btc-500x330" chart="trade-volume"]

* chart is one of:
	* total-bitcoins
	* market-cap
	* transaction-fees
	* transaction-fees-usd
	* network-deficit
	* n-transactions
	* n-transactions-total
	* n-transactions-excluding-popular
	* n-unique-addresses
	* n-transactions-per-block
	* n-orphaned-blocks
	* output-volume
	* estimated-transaction-volume
	* estimated-transaction-volume-usd
	* trade-volume
	* market-price
	* cost-per-transaction-percent
	* cost-per-transaction
	* hash-rate
	* difficulty
	* miners-revenue
	* avg-confirmation-time
	* bitcoin-days-destroyed-cumulative
	* bitcoin-days-destroyed
	* bitcoin-days-destroyed-min-week
	* bitcoin-days-destroyed-min-month
	* bitcoin-days-destroyed-min-year
	* blocks-size
	* my-wallet-transaction-volume
	* my-wallet-n-users
	* my-wallet-n-tx

* style is one of (see styles.json for predefined ones)
	* default	
	* btc-162x150
	* btc-250x165
	* btc-500x330
	* btc-700x450
	* orange-162x150
	* orange-250x165
	* orange-500x330
	* orange-700x450
	* blue-162x150
	* blue-250x165
	* blue-500x330
	* blue-700x450
	* (162x150 indicates the width in pixel)

* there are the following additional styling options (overriding the style):
	* width: pixel
	* height: pixel
	* curve_color: #RRGGBB
	* background_color: #RRGGBB
	* background_gradient: Start:#RRGGBB,End:#RRGGBB
	* title_font: fontfile.ttf,height_in_pixel,#RRGGBB
	* default_font: fontfile.ttf,height_in_pixel,#RRGGBB
	* x-axis-label-rotation: degrees
	* x-axis-date-format: format string for date (M.Y or Y, etc.)
	* graph-area: left,top,right,bottom (all in pixel)
	* label: Data curve legend
	
        #RRGGBB is a hexadecimal color code red, green, blue from 00 to FF

== Installation ==

In order to install and operate this widget the following requirements must be met.

= Requirements: =
  * PHP 5.4+
  * cUrl
  * WordPress 3.5+
  * pChart2.1.4 (included in 3rdparty folder)

== Changelog ==

= 0.4.0 =

* compatible with WordPress 4.2

= 0.3.9 =

* fixed protocol for loading assets

= 0.3.8 =

* fixed installation warning

= 0.3.7 =

* removed warning message

= 0.3.6 =

+ compatibel to offical Wordpress plugin repository

= 0.3.5 =

* fixed labels

= 0.3.0 =

* "title" option is now static text

= 0.2.0 =

* added options to configuration

= 0.1.0 =

initial release

== Frequently Asked Questions ==

= Error "The plugin does not have a valid header" = 

If you receive this error message when trying to activate the plugin directly 
after the installation please navigate back to the plugin list page and activate 
the plugin from there.

== Upgrade Notice ==

Currently we don't have upgrade notices for you.

== Screenshots ==

For screenshots please have a look at the official website (http://www.bitcoin-welt.com/)


CHANGELOG
0.4.0

* compatible with WordPress 4.2

0.3.9

* fixed protocol for loading assets

0.3.8

* fixed installation warning

0.3.7

* removed warning message

0.3.6

* remove wp-load.php include

0.3.0

* "title" option is now static text



php/admin.php
<?php
/**
 * Copyright 2014 Both Interact GmbH (office@both-interact.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
function btccharts_config_page() {
    $styles = array();
    foreach (wp_load_alloptions() as $name => $val) {
        if (substr($name, 0, 16) == "btccharts:style#") {
            $styles[substr($name, 16)] = unserialize($val)->parameters;
        }
    }
    echo "<select size='20' style='height:120pt'>";
    foreach (array_map('htmlspecialchars', array_keys($styles)) as $htm) {
        echo "<option value=\"$htm\">$htm</option>\n";
    }
    echo "</select>";

    foreach ($styles as $name => $val) {
        echo "<div id='$name'>";
        echo "<label for='$name'>" . htmlspecialchars(__("Style name")) . "</label>";
        echo "<input id='style_name_$name' name='style_name[]' value='" . htmlspecialchars($name) . "' />";
        foreach (btccharts_settings_template() as $k => $def_v) {
            $v = @$val[$k];
            $id = htmlspecialchars($name . '_' . str_replace('-', '_', $k));
            $fldName = str_replace('-', '_', $k);
            echo "<div>";
            echo "<label for='$id'>" . htmlspecialchars($k) . "</label>";
            echo "<input id='$id' name='{$fldName}[]' value='" . htmlspecialchars($v) . "' />";
            echo "</div>";
        }
        echo "</div>";
    }
}


php/widget.php
<?php
/**
 * Copyright 2014 Both Interact GmbH (office@both-interact.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
class BtcChartsWidget extends WP_Widget {

    public function __construct() {
        parent::__construct("btcharts", "Bitcoin Charts", array('description' => __('Bitcoin Charts Widget', 'text_domain')));
    }

    public function form($instance) {
        $serviceUrl = "blockchain.info";
        $chartType = "market-price";
        $chartStyle = "default";
        $domain = "";
        $chartTypeOptions = json_decode(file_get_contents(dirname(__FILE__) . "/../js/chart-types.json"), true);
        $chartStyles = json_decode(file_get_contents(dirname(__FILE__) . "/../js/styles.json"), true);

        if ($instance) {
            $this->formData = $instance;
            $serviceUrl = $instance['serviceUrl'];
            $chartType = $instance['chartType'];
            $chartStyle = $instance['style'];
            $domain = $instance['domain'];
        }
        ?>
        <div>
            <div><?php echo __('Service', 'text_domain'); ?></div>
            <select size="1" id="<?php echo $this->get_field_id("serviceUrl"); ?>" name="<?php echo $this->get_field_name("serviceUrl"); ?>">
                <?php
                foreach ($chartTypeOptions as $type => $opt) {
                    $sel = ($type == $serviceUrl) ? " selected=\"selected\"" : "";
                    echo "<option value=\"" . esc_attr($type) . "\"$sel>" . esc_html($opt['label']) . "</option>\n";
                }
                ?>
            </select>
        </div>
        <div>
            <div><?php echo __('Chart', 'text_domain'); ?></div>
            <select size="1" id="<?php echo $this->get_field_id("chartType"); ?>" name="<?php echo $this->get_field_name("chartType"); ?>">
                <?php
                foreach ($chartTypeOptions as $chartService => $data) {
                    foreach ($data['$chart-type'] as $type => $opt) {
                        $sel = ($type == $chartType) ? " selected=\"selected\"" : "";
                        echo "<option value=\"" . esc_attr($type) . "\"$sel>" . esc_html($opt['label']) . "</option>\n";
                    }
                }
                ?>
            </select>
        </div>
        <div>
            <div><?php echo __('Style', 'text_domain'); ?></div>
            <select size="1" id="<?php echo $this->get_field_id("style"); ?>" name="<?php echo $this->get_field_name("style"); ?>">
                <?php
                foreach ($chartStyles as $styleName => $data) {
                    $sel = ($styleName == $chartStyle) ? " selected=\"selected\"" : "";
                    echo "<option value=\"" . esc_attr($styleName) . "\"$sel>" . esc_html($styleName) . "</option>\n";
                }
                ?>
            </select>
        </div>
        <!--<div>
            <div><?php echo __('Domain', 'text_domain'); ?></div>
        <?php $this->input("domain"); ?>
        </div>-->
        <!--<div>
            <div><?php echo __('Title (Diagram)', 'text_domain'); ?></div>
        <?php echo 'bitcoin-welt.com'; /* $this->input("title"); */ ?>
        </div>-->
        <!--<div>
            <div><?php echo __('Label (Data)', 'text_domain'); ?></div>
        <?php $this->input("label"); ?>
        </div>-->
        <!-- <div>
        <div><?php echo __('Width/Height', 'text_domain'); ?></div>
        <?php $this->input("width", 4); ?>/<?php $this->input("height", 4); ?>
        </div>-->
        <?php
    }

    private function input($name, $width = null) {
        $v = @$this->formData[$name];
        echo '<input id="' . $this->get_field_id($name) . '" name="' . $this->get_field_name($name) . '" value="'
        . ($v ? esc_attr($v) : '') . '"';
        if ($width)
            echo ' style="width:' . ($width * 6) . 'pt"';
        echo ' />';
    }

    public function widget($args, $instance) {
        echo "<aside class='widget btccharts_widget'>";
        echo btccharts_render_chart($instance['domain'], $instance['serviceUrl'], $instance['chartType'], $instance);
        echo "</aside>";
    }

}

function btccharts_register_widget() {
    register_widget('BtcChartsWidget');
}

add_action('widgets_init', 'btccharts_register_widget');


php/activate.php
<?php
/**
 *
 * Copyright 2014 Both Interact GmbH (office@both-interact.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
function btccharts_activate() {
    add_option("btccharts:version", $GLOBALS['btccharts_version']);
    // timeout in seconds (for wp transient objects)
    add_option("btccharts:cache_timeout", 3600);
    //make new cache ids after active 
    add_option("btccharts:cache_id", rand());

    $chartTypeOptions = json_decode(file_get_contents(dirname(__FILE__) . "/../js/chart-types.json"), true);
    $id = 0;
    foreach ($chartTypeOptions as $service => $serviceOpts) {
        $types = $serviceOpts['$chart-type'];
        unset($serviceOpts['$chart-type']);
        $o = new btccharts_Option( ++$id, 'service', $service, $serviceOpts);
        add_option("btccharts:service#$service", $o);
        foreach ($types as $chartName => $chartOpts) {
            $o = new btccharts_Option( ++$id, $service, $chartName, $chartOpts);
            add_option("btccharts:$service#$chartName", $o);
        }
    }
    $styles = json_decode(file_get_contents(dirname(__FILE__) . "/../js/styles.json"), true);
    foreach ($styles as $styleName => $styleOpts) {
        $o = new btccharts_Option($id++, 'style', $styleName, $styleOpts);
        add_option("btccharts:style#$styleName", $o);
    }
}

function btccharts_deactive() {
    foreach (wp_load_alloptions() as $name => $val) {
        if (substr($name, 0, 10) == "btccharts:") {
            delete_option($name);
        }
    }
}


php/shortcode.php
<?php
/**
 * Copyright 2014 Both Interact GmbH (office@both-interact.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
add_shortcode('btcchart', 'btccharts_shortcode_btcchart');

function btccharts_shortcode_btcchart($args) {
    $a = array_merge(array(
        "service" => "blockchain.info",
        "chart" => "market-price",
            ), $args);

    return //"hi<pre>".htmlspecialchars(var_export($a, true))."</pre>".
            btccharts_render_chart($a['domain'], $a['service'], $a['chart'], $a);
}


bitcoin-charts-widget.php
<?php
/**
 * Plugin Name: Bitcoin Charts Widget
 * Plugin URI: http://www.bitcoin-welt.com/wordpress-plugin-bitcoin-charts-als-widget/
 * Author: Both Interact GmbH
 * Author URI: http://www.both-interact.com
 * Version: 0.4.0
 * Tags: bitcoin, charts, widget, bitcoin-welt, blockchain.info, service
 * Requires at least: 3.5.0
 * Tested up to: 4.2
 * License: GPL2 or later
 * License URI: http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 * Description: Displays bitcoin charts as widget using data from external 
 *              services such as blockchain.info.
 *
 * Copyright 2014 Both Interact GmbH (office@both-interact.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
defined('ABSPATH') or die("No script kiddies please!");

$btccharts_version = "0.3.9";

require_once 'php/chart.php';
require_once 'php/activate.php';
require_once 'php/widget.php';
require_once 'php/shortcode.php';
require_once 'php/admin.php';

function btccharts_plugin_action_links($links, $file) {
    if ($file == plugin_basename(dirname(__FILE__) . '/bitcoin-charts-widget.php')) {
        // NOT YET DONE $links[] = '<a href="admin.php?page=btccharts-key-config">'.__('Settings').'</a>';
    }
    return $links;
}

function btccharts_admin_menu() {
    /* NOT YET DONE add_submenu_page('plugins.php', 
      __('Bitcoin Charts Configuration'), __('Bitcoin Charts Configuration'),
      'manage_options', 'btccharts-key-config', 'btccharts_config_page'); */
}

// always reload after version change
if (get_option("btccharts:version") != $btccharts_version) {
    btccharts_deactive();
    btccharts_activate();
}

register_activation_hook(__FILE__, "btccharts_activate");
register_deactivation_hook(__FILE__, "btccharts_deactive");
add_filter('plugin_action_links', 'btccharts_plugin_action_links', 10, 2);
add_action('admin_menu', 'btccharts_admin_menu');



php/chart.php
<?php
/**
 * Copyright 2014 Both Interact GmbH (office@both-interact.com)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
$btccharts_show_timings = false;

class btccharts_Option {

    /** @var int $id  technical id */
    public $id;

    /** @var string $class Option classification
     * service 
     * domain
     * style 
     * {service-id}
     */
    public $class;

    /** @var string $code Option indentification (based on classification) */
    public $code;

    /** @var array $parameters Option parameters. An array of values or arrays */
    public $parameters;

    public function __construct($id, $class, $code, $parameters) {
        $this->id = $id;
        $this->class = $class;
        $this->code = $code;
        $this->parameters = $parameters;
    }

}

function btccharts_get_option($class, $code) {
    if ($code == null)
        return get_option("btccharts:$class");
    return get_option("btccharts:$class#$code");
}

function btccharts_get_options($class) {
    $ret = array();
    foreach (wp_load_alloptions() as $name => $opt) {
        if (substr($name, 0, 10) == 'btccharts:') {
            if ($opt->class == $class)
                $ret[] = $opt;
        }
    }
    return $ret;
}

function btccharts_cache($uid, $file) {
    set_transient("btccharts:$uid", file_get_contents($file), 60 * 60 * 24);
}

function btccharts_get_cached($uid) {
    get_transient("btccharts:$uid");
}

function btccharts_get_url($url) {
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_TIMEOUT, 30);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
    $result = curl_exec($ch);
    curl_close($ch);
    return $result;
}

function btccharts_settings_template() {
    return array(
        "width" => 400,
        "height" => 300,
        "title" => "",
        "label" => "",
        "image-href" => "",
        "curve_color" => "#FF0000",
        "background_color" => "#f7931a,Dash:#f7931a",
        "background_gradient" => "Start:#f7931a,End:#D88116",
        "title_font" => "calibri.ttf,8,#FFF",
        "default_font" => "verdana.ttf,6,#000",
        "x-axis-label" => "",
        "x-axis-label-rotation" => 40,
        "x-zoom" => 1.0,
        "x-axis-date-format" => "M.Y",
        "y-axis-label" => "",
        "y-zoom" => 1.0,
        "label-pos" => "-100,9", // left, top
        "title-pos" => "10,16", // left, top
        "graph-area" => "70,30,-10,-60" // left, top, right, bottom; -x .. distance from bottom/right border
    );
}

function btccharts_filter_settings($settings) {
    $template = btccharts_settings_template();
    $filtered = array();
    // filter out everything not in template
    foreach ($settings as $k => $v) {
        if (key_exists($k, $template) && !empty($v)) {
            if ($k == "width" && (0 + $v < 120 || 0 + $v > 2000)) {
                continue;
            } elseif ($k == "height" && (0 + $v < 150 || 0 + $v > 2000)) {
                continue;
            } elseif ($k == "x-axis-label-rotation" && (0 + $v < 0 || 0 + $v > 360)) {
                continue;
            }
            $filtered[$k] = $v;
        }
    }
    return $filtered;
}

/**
 * 
 * @param array $settings
 * @return array style attributes
 */
function btccharts_style($settings) {

    if (key_exists('style', $settings)) {
        $style = btccharts_get_option('style', $settings['style']);
        if (is_object($style))
            $default = $style->parameters;
    }
    if (!isset($default) || !is_array($default) || count($default) == 0) {
        $default = btccharts_settings_template();
    }
    foreach (btccharts_filter_settings($settings) as $k => $v) {
        $default[$k] = $v;
    }
    return $default;
}

/** Create a HTML snippet with a link to the chart image.
 * 
 * @param unknown $serviceId
 * @param unknown $chartType
 * @param unknown $settings
 * @return string
 */
function btccharts_render_chart($domain, $serviceId, $chartType, $settings) {
    $service = btccharts_get_option("service", $serviceId)->parameters;
    $opts = btccharts_get_option($serviceId, $chartType)->parameters;
    $html = "";
    $html .= "<div class='btcchart'>";
    $s = btccharts_style(array_merge($opts, $settings));
    $imgLink = $s['image-href'] ? $s['image-href'] : $service['image-href'];

    $imgUrl = btccharts_render_chart_image(false, $domain, $serviceId, $chartType, $s);

    $html .= "<div>";
    if ($imgLink)
        $html .= '<a href="' . htmlspecialchars($imgLink) . '">';
    $html .= "<img src=\"$imgUrl\"/>";
    if ($imgLink)
        $html .= "</a>";
    $html .= "</div>";
    $html .= "</div>";
    return $html;
}

function btccharts_render_chart_image($writeToStdout, $domain, $serviceUrl, $chartType, $settings = array()) {
    $t0 = microtime(true);
    if ($writeToStdout)
        ob_start();
    //$chartTypeOptions = json_decode(file_get_contents(dirname(__FILE__)."/../js/chart-types.json"), true);
    //$serviceDesc = $chartTypeOptions[$serviceUrl];
    //$opts = $serviceDesc['$chart-type'][$chartType];

    $cache_id = btccharts_get_option("cache_id", "");
    $cache_timeout = btccharts_get_option("cache_timeout", 60);
    $serviceDesc = btccharts_get_option("service", $serviceUrl)->parameters;
    $opts = btccharts_get_option($serviceUrl, $chartType)->parameters;
    $service = str_replace('$chart-type', $chartType, $serviceDesc['service-url']);
    $s = btccharts_style(array_merge($opts, $settings));
    $id = md5($domain . $chartType . serialize($s) . serialize($serviceDesc) . $cache_id);
    //user_error("btccharts_render_chart_image id ($id)", E_USER_WARNING);
    $cached = get_transient("btccharts:$id");
    if (!is_null($cached) && strlen($cached) > 0) {
        if ($writeToStdout) {
            $img = base64_decode($cached);
            if ($GLOBALS['btccharts_show_timings'])
                user_error("btccharts_render_chart_image from cache: t0=" . (microtime(true) - $t0), E_USER_WARNING);
            ob_end_clean();
            header("Expires: " . date('r', time() + $cache_timeout));
            header("Content-Type: image/png");

            echo($img);
            return;
        } else {
            if ($GLOBALS['btccharts_show_timings'])
                user_error("btccharts_render_chart_image from cache: t0=" . (microtime(true) - $t0), E_USER_WARNING);
            return "data:image/png;base64," . $cached;
        }
    }

    $t1 = microtime(true);
    if ($GLOBALS['btccharts_show_timings'])
        user_error("btccharts_render_chart_image init: t=" . ($t1 - $t0), E_USER_WARNING);
    $data = json_decode(btccharts_get_url($service));
    $t2 = microtime(true);
    if ($GLOBALS['btccharts_show_timings'])
        user_error("btccharts_render_chart_image service: t=" . ($t2 - $t1), E_USER_WARNING);
    /* CAT:Line chart */

    /* pChart library inclusions */
    $pChartRoot = dirname(__FILE__) . '/../3rdparty/pChart/';
    require_once $pChartRoot . 'class/pData.class.php';
    require_once $pChartRoot . 'class/pDraw.class.php';
    require_once $pChartRoot . 'class/pImage.class.php';

    $x_zoom = @$opts['x-zoom'] ? $opts['x-zoom'] : 1.0;
    $y_zoom = $s['y-zoom'];
    $date_axis = array();
    $value_axis = array();
    foreach ($data->values as $pt) {
        $date_axis[] = $pt->x / $x_zoom;
        $value_axis[] = $pt->y / $y_zoom;
    }
    /* Create and populate the pData object */
    $MyData = new pData();
    $MyData->addPoints($value_axis, $s['label']);
    $MyData->addPoints($date_axis, "Dates");
    $MyData->setSerieWeight($s['label'], 0);
    $MyData->setSerieShape($s['label'], 0);
    $MyData->setAxisName(0, $s['y-axis-label']);
    $MyData->Data["Series"][$s['label']]["Color"] = rgb($s['curve_color']);
    $MyData->setAbscissaName($s['x-axis-label']);
    $MyData->setAbscissa("Dates");
    $MyData->setXAxisDisplay(AXIS_FORMAT_TIME, $s["x-axis-date-format"]);

    /* Create the pChart object */
    $myPicture = new pImage($s['width'], $s['height'], $MyData);

    /* Turn of Antialiasing */
    $myPicture->Antialias = TRUE;

    /* Draw the background */
    //f7931a 247 147 26
    $Settings = rgb($s['background_color'], array("Dash" => 1), 50);
    $myPicture->drawFilledRectangle(0, 0, $s['width'], $s['height'], $Settings);

    /* Overlay with a gradient */
    $Settings = rgb($s['background_gradient'], array(), 50);
    $myPicture->drawGradientArea(0, 0, $s['width'], $s['height'], DIRECTION_VERTICAL, $Settings);
    $myPicture->drawGradientArea(0, 0, $s['width'], 20, DIRECTION_VERTICAL, rgb("Start:#000,End:#323232", array("Alpha" => 80)));

    /* Add a border to the picture */
    $myPicture->drawRectangle(0, 0, $s['width'] - 1, $s['height'] - 1, rgb("#000"));

    /* Write the chart title */
    $title = empty($s['title']) ? $serviceDesc['label'] : $s['title'];
    if (trim($title) != '') {
        $myPicture->setFontProperties(font($s['title_font']));
        $g = btccharts_pos('title-pos', $s);
        $myPicture->drawText($g[0], $g[1], $title, array("Align" => TEXT_ALIGN_BOTTOMLEFT));
    }
    /* Set the default font */
    $myPicture->setFontProperties(font($s['default_font']));

    /* Define the chart area */
    // $s['graph-area'] == "<left>,<top>,<right>,<bottom>"
    $g = array_map('trim', explode(',', $s['graph-area']));
    while ($g[0] < 0)
        $g[0] += $s['width'];
    while ($g[1] < 0)
        $g[1] += $s['height'];
    while ($g[2] < 0)
        $g[2] += $s['width'];
    while ($g[3] < 0)
        $g[3] += $s['height'];
    $myPicture->setGraphArea($g[0], $g[1], $g[2], $g[3]);

    /* Draw the scale */
    $scaleSettings = rgb("Grid:#C8C8C8", array("XMargin" => 10, "YMargin" => 10, "Floating" => TRUE, "RemoveSkippedAxis" => TRUE,
        "ManualScale" => array(0 => array("Min" => 10, "Max" => 100)),
        "DrawSubTicks" => FALSE, "CycleBackground" => TRUE,
        "LabelingMethod" => LABELING_DIFFERENT,
        "LabelRotation" => $s['x-axis-label-rotation']));
    //$scaleSettings = array("XMargin"=>10,"YMargin"=>10,"Floating"=>TRUE,"GridR"=>200,"GridG"=>200,"GridB"=>200,"RemoveSkippedAxis"=>TRUE,"DrawSubTicks"=>FALSE,"Mode"=>SCALE_MODE_START0,"LabelingMethod"=>LABELING_DIFFERENT);
    $myPicture->drawScale($scaleSettings);

    /* Turn on Antialiasing */
    $myPicture->Antialias = TRUE;

    /* Enable shadow computing */
    $myPicture->setShadow(TRUE, array("X" => 1, "Y" => 1, "R" => 0, "G" => 0, "B" => 0, "Alpha" => 10));

    /* Draw the line chart */
    $myPicture->drawLineChart();
    $myPicture->drawPlotChart(array(
        "DisplayValues" => FALSE, "PlotBorder" => TRUE, "BorderSize" => 2,
        "Surrounding" => -60, "BorderAlpha" => 80));

    /* Write the chart legend */
    $g = btccharts_pos('label-pos', $s);
    $myPicture->drawLegend($g[0], $g[1], rgb("Font:#FFF", array(
        "Style" => LEGEND_NOBORDER, "Mode" => LEGEND_HORIZONTAL)));

    /* Render the picture (choose the best way) */
    $fn = tempnam(sys_get_temp_dir(), 'btccharts');
    @$myPicture->render($fn);
    $img = file_get_contents($fn);
    if ($GLOBALS['btccharts_show_timings'])
        user_error("btccharts_render_chart_image image render: t=" . (microtime(true) - $t2), E_USER_WARNING);
    set_transient("btccharts:$id", base64_encode($img), $cache_timeout);
    if (!$writeToStdout) {
        $ret = "data:image/png;base64," . base64_encode($img);
    } else {
        unlink($fn);
        ob_end_clean();
        header("Expires: " . date('r', time() + $cache_timeout));
        header("Content-Type: image/png");
        echo $img;
        $ret = null;
    }
    $t3 = microtime(true);
    if ($GLOBALS['btccharts_show_timings'])
        user_error("btccharts_render_chart_image image: t=" . ($t3 - $t2), E_USER_WARNING);
    return $ret;
}

function btccharts_pos($field, &$settings) {
    $g = array_map('trim', explode(',', $settings[$field]));
    for ($i = 0; $i + 1 < count($g); $i += 2) {
        while ($g[$i + 0] < 0)
            $g[$i + 0] += $settings['width'];
        while ($g[$i + 1] < 0)
            $g[$i + 1] += $settings['height'];
    }
    return $g;
}

function rgb($webs, $dest = array(), $def_alpha = 100) {
    foreach (array_map('trim', explode(',', $webs)) as $web) {
        $prefix = "";
        if (strpos($web, ':') !== false)
            list($prefix, $web) = array_map('trim', explode(':', $web));
        if ($prefix == "Alpha") {
            $dest['Alpha'] = (int) $web;
        }
        $ofs = strlen($web) % 3;
        $len = (int) (strlen($web) / 3);
        foreach (array("R", "G", "B") as $i => $c) {
            $dest[$prefix . $c] = hexdec(substr($web, $ofs + $i * $len, $len)) * ($len == 1 ? 16 : 1);
        }
    }
    if (!key_exists('Alpha', $dest))
        $dest['Alpha'] = $def_alpha;

    return $dest;
}

function font($spec, $dest = array()) {
    list($filename, $pixelSize, $color) = explode(",", $spec);
    $fontDir = dirname(__FILE__) . '/../3rdparty/pChart/fonts';
    return rgb($color, array_merge($dest, array("FontName" => "$fontDir/$filename", "FontSize" => $pixelSize)));
}
